README.md

# Ubiquitin Analysis

The goal of Ubiquitin Analysis is to analysis the relationships between two different ubiquitin proteins.

## Installation
To install the latest version of the package:
``` r
require("devtools")
install_github("sealigu/UbiquitinAnalysis")
library("UbiquitinAnalysis")
```
To run the shinyApp:
```r
UbiquitinAnalysis::runUA()
```

## Overview
```r
browseVignettes("UbiquitinAnalysis")
```
This package contains 4 functions to demonstrate components of a simple R package.

### Function 1: PlotResModification()
```r
UbiquitinAnalysis::PlotResModification("Q9UHB7", "Q9UKV5")
```
![](./inst/pic/Res._Modification_Example.png)

Input two selected protein uniprot id (e.g. "Q9UHB7" and "Q9UKV5"), the output is a plot to show the amino acid modification position in the selected protein sequences. The output will help to compare whether these two proteins have the same modified position in both sequences.

### Function 2: PlotProteinInteractions()
```r
UbiquitinAnalysis::PlotProteinInteractions("Q9UHB7", "Q9UKV5")
```
![](./inst/pic/Protein_Interaction_Example.png)

Input two selected protein uniprot id (e.g. "Q9UHB7" and "Q9UKV5"), the output is a plot to show the interacting proteins of the selected proteins. The output will help to compare whether these two proteins have the same interacting protein.

### Function 3: GetSimilarPercentage()
```r
UbiquitinAnalysis::GetSimilarPercentage("Q9UHB7", "Q9UKV5")
```
![](./inst/pic/Similar_Percentage_Example.png)

Input two selected protein uniprot id (e.g. "Q9UHB7" and "Q9UKV5"), the output is a percentage number that shows the similarity of the two selected protein sequences.

# One-minute pitch slide
![](./inst/pic/TANG_S_A1.png)

## Example
``` r
library(Ubiquitin Analysis);
lsf.str("package:UbiquitinAnalysis");

PlotResModification("Q9UHB7", "Q9UKV5");
PlotProteinInteractions("Q9UHB7", "Q9UKV5");
GetSimilarPercentage("Q9UHB7", "Q9UKV5");
```
- PlotResModification("Q9UHB7", "Q9UKV5")
- PlotProteinInteractions("Q9UHB7", "Q9UKV5")

The functions PlotResModification and PlotProteinInteractions were authored by Shiyun. Part of the code is used from StackOverFlow. The link and the authors are given in the code.

- GetSimilarPercentage("Q9UHB7", "Q9UKV5")

The function GetSimilarPercentage was authored by Shiyun. The Levenshtein distance algorithm was used to measuring the differences between two sequences.

- library(UniprotR)

The library UniprotR was used to retrieve data from UniProt services.

## References
1. Package UniprotR: Retrieving Information of Proteins from Uniprot 
https://cran.r-project.org/web/packages/UniprotR/index.html


NEWS.md
# UbiquitinAnalysis 0.1.0
* Final version for UbiquitinAnalysis R package


test/testthat.R
library(testthat)
library("UbiquitinAnalysis")

test_check("UbiquitinAnalysis")



test/test_mainfuncs.R
library(testthat)
library("UbiquitinAnalysis")

test_that("To test GetSimilarPercentage", {
  protein1 = "Q9UHB7"
  protein2 = "Q9UKV5"

  value <- GetSimilarPercentage(protein1, protein2)

  expect_that(value, is_a("percentage"))
  expect_that(trunc(value), equals("56.8%"))
})

test_check("UbiquitinAnalysis")



test/testthat/test_GetSimPercen.R
context("Get similar percentage number of two proteins")
library("UbiquitinAnalysis")

test_that("To test GetSimilarPercentage", {
  protein1 = "Q9UHB7"
  protein2 = "Q9UKV5"

  value <- GetSimilarPercentage(protein1, protein2)

  expect_that(value, is_a("percentage"))
  expect_that(trunc(value), equals("56.8%"))
})


R/GetSimPercen.R
#' This function is used to calculate the similarities between
#' two protein sequences with a percentage output.
#'
#' @param  protein1 A protein Uniprot
#' @param  protein2 A protein Uniprot
#'
#' @return Return a percentage number tells the
#' similarities between selected protein sequences
#' @examples
#' GetSimilarPercentage("Q9UHB7", "Q9UKV5")
#'
#' @name GetSimPercen
#' @rdname GetSimPercen
#' @import UniprotR
#' @import stringi
#' @import scales
#' @import pheatmap
#'

citation("UniprotR");
citation("stringi");
citation("scales");

if(!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

if(!requireNamespace("UniprotR", quietly = TRUE)) {
  install.packages("UniprotR")
}

if(!requireNamespace("stringr", quietly = TRUE)) {
  install.packages("stringr")
}

if(!requireNamespace("scales", quietly = TRUE)) {
  install.packages("scales")
}

if(!requireNamespace("pheatmap", quietly = TRUE)) {
  install.packages("pheatmap")
}

library(UniprotR);
library(stringr);
library(scales);
library(pheatmap);

GetSimilarPercentage <- function(protein1 = "Q9UHB7", protein2 = "Q9UKV5"){
  protein1 <- toString(protein1)
  protein2 <- toString(protein2)
  pro1_seq <- as.character(UniprotR::GetSequences(protein1)$Sequence);
  pro2_seq <- as.character(UniprotR::GetSequences(protein2)$Sequence);

  if(identical(pro1_seq, "NA") | identical(pro2_seq, "NA")){
    return("One of the protein sequence is N/A.");
  }

  #By using the Levenshtein distance algorithm
  #for measuring the difference between sequences

  pro1len <- nchar(pro1_seq);
  pro2len <- nchar(pro2_seq);

  #check the sequence length is not 0
  if(identical(pro1len, 0)){
    return(pro2len);
  }
  if(identical(pro2len,0)){
    return(pro1len);
  }

  #create a new empty matrix
  tab <- matrix(0, nrow = pro1len+1, ncol = pro2len+1);
  #initial first row of the matrix
  for(i in 1:(pro1len+1)){
    tab[i, 1] <- i;
  }
  #initial first col of the matrix
  for(j in 1:(pro2len+1)){
    tab[1, j] <- j;
  }

  for(i in 2:pro1len+1){
    ch1 <- substr(pro1_seq, i-1, i-1);
    for(j in 2:pro2len+1){
      ch2 <- substr(pro2_seq, j-1, j-1);
      if(identical(ch1, ch2)){
        hit = 0;
      }else{
        hit = 1;
      }

      tab[i,j] <- min(tab[i-1, j]+1, tab[i, j-1]+1, tab[i-1, j-1]+hit);
    }
  }

  tab_val <- tab[pro1len+1, pro2len+1];
  #using percent() function from library(scales) to convert decimal to percentage
  v <- scales::percent(1-tab_val/max(pro1len, pro2len))
  return(v);
}



R/PlotProInter.R
#' This function is used to plot a graph that shows all the
#' interacted proteins with the two input proteins. The output is
#' a graph to show all the interactors and comparing to see whether
#' they have the same interacting proteins.
#'
#' @param  protein1 A protein Uniprot
#' @param  protein2 A protein Uniprot
#'
#' @return Returns a plot to show all the protein interactors
#' @examples
#' PlotProteinInteractions("Q9UHB7", "Q9UKV5")
#'
#' @name PlotProInter
#' @import UniprotR
#' @import stringi
#' @import scales
#' @import pheatmap
#'

citation("UniprotR");
citation("stringi");
citation("scales");
citation("pheatmap");

if(!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

if(!requireNamespace("UniprotR", quietly = TRUE)) {
  install.packages("UniprotR")
}

if(!requireNamespace("stringr", quietly = TRUE)) {
  install.packages("stringr")
}

if(!requireNamespace("scales", quietly = TRUE)) {
  install.packages("scales")
}

if(!requireNamespace("pheatmap", quietly = TRUE)) {
  install.packages("pheatmap")
}

library(UniprotR);
library(stringr);
library(scales);
library(pheatmap);

PlotProteinInteractions <- function(protein1 = "Q9UHB7", protein2 = "Q9UKV5") {
  protein1 <- toString(protein1)
  protein2 <- toString(protein2)
  pro1_obj <- as.character(UniprotR::GetProteinInteractions(protein1)$Interacts.with);
  pro2_obj <- as.character(UniprotR::GetProteinInteractions(protein2)$Interacts.with);

  pro1_list <- strsplit(pro1_obj, ";"); #list
  pro2_list <- strsplit(pro2_obj, ";");
  pro1_sp <- unlist(pro1_list); #character
  pro2_sp <- unlist(pro2_list);

  #change "Itself" into its uniprot id
  for(i in 1:length(pro1_sp)){
    if(identical(pro1_sp[i], "Itself")){
      pro1_sp[i] <- as.character(protein1);
    }
  }

  for(j in 1:length(pro2_sp)){
    if(identical(pro2_sp[j], "Itself")){
      pro2_sp[j] <- as.character(protein2);
    }
  }

  pro1_mat <- matrix(1, nrow =  1,
                     ncol = length(pro1_sp),
                     dimnames = list(as.character(protein1), c(pro1_sp)));
  pro2_mat <- matrix(1, nrow = 1,
                     ncol = length(pro2_sp),
                     dimnames = list(as.character(protein2), c(pro2_sp)));

  # https://stackoverflow.com/questions/5738773/r-how-to-merge-two-matrix-according-to-their-column-and-row-names
  # authors: Chase and Joris Meys
  # According to above code to merge two matrix together
  # and convert list to matrix

  df <- merge(pro1_mat, pro2_mat, by = "row.names", all = TRUE);
  df[is.na(df)] <- 0; #replacing NA's with 0
  tab <- as.matrix(df[-1]);
  rownames(tab) <- df[,1];

  col<- colorRampPalette(c("white", "blue"))(256);
  final_tab <- pheatmap::pheatmap(tab, scale = "none", col = col,
                        cluster_cols = FALSE, cluster_rows = FALSE);
  print(typeof(tab))
  return(tab);
}


R/PlotResModif.R
#' This function is used to plot a graph that shows all the
#' residue modification positons in the selected protein sequences.
#' Input two interested proteins, the plot will show the modified positon
#' for each selected protein and comparing to see whether these two proteins
#' have the same modified position
#'
#' @param  protein1 A protein Uniprot
#' @param  protein2 A protein Uniprot
#'
#' @return Returns a plot to show all the protein residue positions
#' @examples
#' PlotResModification("Q9UHB7", "Q9UKV5")
#'
#' @name PlotResModif
#' @import UniprotR
#' @import stringi
#' @import scales
#' @import pheatmap
#'


citation("UniprotR");
citation("stringi");
citation("scales");
citation("pheatmap");

if(!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

if(!requireNamespace("UniprotR", quietly = TRUE)) {
  install.packages("UniprotR")
}

if(!requireNamespace("stringr", quietly = TRUE)) {
  install.packages("stringr")
}

if(!requireNamespace("scales", quietly = TRUE)) {
  install.packages("scales")
}

if(!requireNamespace("pheatmap", quietly = TRUE)) {
  install.packages("pheatmap")
}

library(UniprotR);
library(stringr);
library(scales);
library(pheatmap);

PlotResModification <- function(protein1 = "Q9UHB7", protein2 = "Q9UKV5") {
  protein1 <- toString(protein1)
  protein2 <- toString(protein2)
  pro1_r <- as.character(UniprotR::GetPTM_Processing(protein1)$Modified.residue);
  pro1_res <- stringr::str_extract_all(pro1_r, stringr::regex("\\d+\\s[A-Za-z]."));
  pro1_res <- unlist(pro1_res);

  pro2_r <- as.character(UniprotR::GetPTM_Processing(protein2)$Modified.residue);
  pro2_res <- stringr::str_extract_all(pro2_r, stringr::regex("\\d+\\s[A-Za-z]."));
  pro2_res <- unlist(pro2_res);

  pro1_res_mat <- matrix(1, nrow = 1,
                         ncol = length(pro1_res),
                         dimnames = list(as.character(protein1), c(pro1_res)));
  pro2_res_mat <- matrix(1, nrow = 1,
                         ncol = length(pro2_res),
                         dimnames = list(as.character(protein2), c(pro2_res)));

  # https://stackoverflow.com/questions/5738773/r-how-to-merge-two-matrix-according-to-their-column-and-row-names
  # authors: Chase and Joris Meys
  # According to above code to merge two matrix together
  # and convert list to matrix

  tab <- merge(pro1_res_mat, pro2_res_mat, by = "row.names", all = TRUE);
  tab[is.na(tab)] <- 0;

  df <- as.matrix(tab[-1]);
  rownames(df) <- tab[,1];

  col<- colorRampPalette(c("white", "purple"))(256);
  f <- pheatmap::pheatmap(df, scale = "none", col = col,
                cluster_cols = FALSE, cluster_rows = FALSE);
  return(df);
}









inst/shiny-scripts/app.R
library(shiny)
library(d3heatmap)

ui <- fluidPage(
  titlePanel("UbiquitinAnalysis: R package for Plotting Protein Relationship"),

  sidebarLayout(
    sidebarPanel(
          textInput(inputId = "UP1",
                  label = h3("Input an UniProt ID"),
                  placeholder = NULL),
          textInput(inputId = "UP2",
                  label = h3("Input an UniProt ID"),
                  placeholder = NULL),
          submitButton("Submit")
      ),

    mainPanel(
      d3heatmapOutput("plot1", width = "100%", height = "400px"),
      d3heatmapOutput("plot2", width = "100%", height = "400px"),
      verbatimTextOutput("percentage")
    )
  )
)

server <- function(input, output) {

  output$plot1 <- renderD3heatmap({
    validate({
      need(input$UP1, 'Check there is at least one letter!')
      need(input$UP2, 'Check there is at least one letter!')
    })
    p1 <- UbiquitinAnalysis::PlotProteinInteractions(input$UP1, input$UP2)
    d3heatmap(p1, dendrogram = "none", colors = c("white", "lightblue"))
  })

  output$plot2 <- renderD3heatmap({
    validate({
      need(input$UP1, 'Check there is at least one letter!')
      need(input$UP2, 'Check there is at least one letter!')
    })
    p2 <- UbiquitinAnalysis::PlotResModification(input$UP1, input$UP2)
    d3heatmap(p2, dendrogram = "none", colors = c("white", "gray88"))
  })

  output$percentage <- renderText({
    validate({
      need(input$UP1, 'Check there is at least one letter!')
      need(input$UP2, 'Check there is at least one letter!')
    })

    p3 <- UbiquitinAnalysis::GetSimilarPercentage(input$UP1, input$UP2)
    paste("Protein Sequences Similar Percentage: ", p3)
  })
}

shinyApp(ui = ui, server = server)


R/main.R
# This .R file is used to analyze the ubiquitin proteins
#
# The first function is used to plot a graph that shows all the
# residue modification positons in the selected protein sequences.
# Input two interested proteins, the plot will show the modified positon
# for each selected protein and comparing to see whether these two proteins
# have the same modified position
#
# The second function is used to plot a graph that shows all the
# interacted proteins with the two input proteins. The output is
# a graph to show all the interactors and comparing to see whether
# they have the same interacting proteins.
#
# The third function is used to calculate the similarities between
# two protein sequences with a percentage output.
#
# @param protein1 is the selected protein uniprot id, example "Q86Y01"
# @param protein2 is the selected protein uniprot id, example "Q9NYG5"
#
#

citation("UniprotR");
citation("stringi");
citation("scales");
citation("pheatmap");
library(UniprotR);
library(stringr);
library(scales);
library(pheatmap);

#################################################################
# @examples
# PlotResModification("Q9UHB7", "Q9UKV5")
#
# @return Returns a plot to show the protein modified position

PlotResModification <- function(protein1, protein2) {
  pro1_r <- as.character(GetPTM_Processsing(protein1)$Modified.residue);
  pro1_res <- str_extract_all(pro1_r, regex("\\d+\\s[A-Za-z]."));
  pro1_res <- unlist(pro1_res);

  pro2_r <- as.character(GetPTM_Processsing(protein2)$Modified.residue);
  pro2_res <- str_extract_all(pro2_r, regex("\\d+\\s[A-Za-z]."));
  pro2_res <- unlist(pro2_res);

  pro1_res_mat <- matrix(1, nrow = 1,
                         ncol = length(pro1_res),
                         dimnames = list(as.character(protein1), c(pro1_res)));
  pro2_res_mat <- matrix(1, nrow = 1,
                         ncol = length(pro2_res),
                         dimnames = list(as.character(protein2), c(pro2_res)));

  # https://stackoverflow.com/questions/5738773/r-how-to-merge-two-matrix-according-to-their-column-and-row-names
  # authors: Chase and Joris Meys
  # According to above code to merge two matrix together
  # and convert list to matrix

  tab <- merge(pro1_res_mat, pro2_res_mat, by = "row.names", all = TRUE);
  tab[is.na(tab)] <- 0;

  df <- as.matrix(tab[-1]);
  rownames(df) <- tab[,1];

  col<- colorRampPalette(c("white", "purple"))(256);
  f <- pheatmap(df, scale = "none", col = col);
  return();
}


#################################################################
# @examples
# PlotProteinInteractions("Q9UHB7", "Q9UKV5")
#
# @return Returns a plot to show all the protein interactors

PlotProteinInteractions <- function(protein1, protein2) {
  pro1_obj <- as.character(GetProteinInteractions(protein1)$Interacts.with);
  pro2_obj <- as.character(GetProteinInteractions(protein2)$Interacts.with);

  pro1_list <- strsplit(pro1_obj, ";"); #list
  pro2_list <- strsplit(pro2_obj, ";");
  pro1_sp <- unlist(pro1_list); #character
  pro2_sp <- unlist(pro2_list);

  #change "Itself" into its uniprot id
  for(i in 1:length(pro1_sp)){
    if(identical(pro1_sp[i], "Itself")){
      pro1_sp[i] <- as.character(protein1);
    }
  }

  for(j in 1:length(pro2_sp)){
    if(identical(pro2_sp[j], "Itself")){
      pro2_sp[j] <- as.character(protein2);
    }
  }

  pro1_mat <- matrix(1, nrow =  1,
                     ncol = length(pro1_sp),
                     dimnames = list(as.character(protein1), c(pro1_sp)));
  pro2_mat <- matrix(1, nrow = 1,
                     ncol = length(pro2_sp),
                     dimnames = list(as.character(protein2), c(pro2_sp)));

  # https://stackoverflow.com/questions/5738773/r-how-to-merge-two-matrix-according-to-their-column-and-row-names
  # authors: Chase and Joris Meys
  # According to above code to merge two matrix together
  # and convert list to matrix

  df <- merge(pro1_mat, pro2_mat, by = "row.names", all = TRUE);
  df[is.na(df)] <- 0; #replacing NA's with 0
  tab <- as.matrix(df[-1]);
  rownames(tab) <- df[,1];

  col<- colorRampPalette(c("white", "blue"))(256);
  final_tab <- pheatmap(tab, scale = "none", col = col);
  return(final_tab);
}


##############################################################
# @examples
# GetSimilarPercentage("Q9UHB7", "Q9UKV5");
#
# @return Return a percentage number tells the
# similarities between selected protein sequences

GetSimilarPercentage <- function(protein1, protein2){
  pro1_seq <- as.character(GetSequences(protein1)$Sequence);
  pro2_seq <- as.character(GetSequences(protein2)$Sequence);

  if(identical(pro1_seq, "NA") | identical(pro2_seq, "NA")){
    return("One of the protein sequence is N/A.");
  }

  #By using the Levenshtein distance algorithm
  #for measuring the difference between sequences

  pro1len <- nchar(pro1_seq);
  pro2len <- nchar(pro2_seq);

  #check the sequence length is not 0
  if(identical(pro1len, 0)){
    return(pro2len);
  }
  if(identical(pro2len,0)){
    return(pro1len);
  }

  #create a new empty matrix
  tab <- matrix(0, nrow = pro1len+1, ncol = pro2len+1);
  #initial first row of the matrix
  for(i in 1:(pro1len+1)){
    tab[i, 1] <- i;
  }
  #initial first col of the matrix
  for(j in 1:(pro2len+1)){
    tab[1, j] <- j;
  }

  for(i in 2:pro1len+1){
    ch1 <- substr(pro1_seq, i-1, i-1);
    for(j in 2:pro2len+1){
      ch2 <- substr(pro2_seq, j-1, j-1);
      if(identical(ch1, ch2)){
        hit = 0;
      }else{
        hit = 1;
      }

      tab[i,j] <- min(tab[i-1, j]+1, tab[i, j-1]+1, tab[i-1, j-1]+hit);
    }
  }

  tab_val <- tab[pro1len+1, pro2len+1];
  #using percent() function from library(scales) to convert decimal to percentage
  return(percent(1-tab_val/max(pro1len, pro2len)));
}


